### PR TITLE
LPS-195690 Customer - License Selection for DXP Apps - Trial

### PR DESCRIPTION
**Pull request by ocesar9**
https://liferay.atlassian.net/browse/LPS-196887

### CODE REVIEW

**Comment by ocesar9** [comment]
ci:test:sf

**Comment by ocesar9** [comment]
ci:test:sf

**Comment by ocesar9** [comment]
ci:test:sf

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+						<ClayButton
+							aria-label=""
+							className="align-items-center d-flex justify-content-center license__card__buttons p-2"
+							displayType="primary"
+							onClick={() => {
+								window.location.href = origin;
+							}}
+						>
+							<ClayIcon
+								aria-label="123"
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx:47]: Please use a descriptive aria-label

```suggestion
								aria-label="Horizontal Line"
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+					<div>
+						<p className="license__card__text mb-0">
+							{licenseDescription}
+						</p>
+					</div>
+				</span>
+				<div className="align-items-center d-flex license__card__buttons__container p-1">
+					<span>
+						<ClayButton
+							aria-label=""
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx:39]: Let this empty or add a descriptive aria label

```suggestion

```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+					</span>
+				</div>
+			</div>
+
+			<div className="d-flex flex-column license__card__tier mt-4 p-4">
+				<div className="font-weight-bold license__card__tier__title mb-1">
+					License Prices
+				</div>
+
+				{licensetiers.map((license: any, index: any) => (
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseCard/index.tsx:80]:
```suggestion
				{licensetiers.map((license, index) => (
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {useCallback, useEffect, useState} from 'react';
+
+import {CardButton} from '../../../../components/CardButton/CardButton';
+import {GetClayIcon} from '../../../../components/GetClayIcon/GetClayIcon';
+import {PaidTimeline} from '../PaidTimeline/index';
+import {TrialTimeline} from '../TrialTimeline/index';
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:11]:
```suggestion
import {PaidTimeline} from '../PaidTimeline';
import {TrialTimeline} from '../TrialTimeline';
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	const [trialSKU, setTrialSKU] = useState<SKU>();
+
+	const hasTrialSkuVerification = useCallback(() => {
+		selectedProduct?.skus?.some((sku) => {
+			sku.skuOptions.some((options) => {
+				if (options.key === 'trial' && options.value === 'yes') {
+					setTrialSKU(sku);
+				}
+			});
+		});
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:34]: Wrong use of `some`

```suggestion
		selectedProduct?.skus?.forEach((sku) => {
			sku.skuOptions.forEach((options) => {
				if (options.key === 'trial' && options.value === 'yes') {
					setTrialSKU(sku);
				}
			});
		});
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+				<CardButton
+					description="Try now. Pay Later"
+					disabled={trialSKU ? false : true}
+					icon={
+						<GetClayIcon
+							className="license-icon"
+							icon="check-circle"
+						/>
+					}
+					onClick={() => handleTimelineSelect('trial')}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:63]:
```suggestion
					onClick={() => setSelectedTimeline('trial')}
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			onSelectLicense(true, trialSKU);
+		}
+	};
+
+	return (
+		<div className="license-selector-timeline">
+			<div className="license-selector mb-6">
+				<CardButton
+					description="Try now. Pay Later"
+					disabled={trialSKU ? false : true}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:56]:
```suggestion
					disabled={!trialSKU}
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+					description="Pay Today"
+					disabled={false}
+					icon={
+						<GetClayIcon
+							className="license-icon"
+							icon="credit-card"
+						/>
+					}
+					onClick={() => handleTimelineSelect('paid')}
+					selected={selectedTimeline === 'paid' ? true : false}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:79]:
```suggestion
					selected={selectedTimeline === 'paid'}
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+			{selectedTimeline ? (
+				<div className="timeline-container">
+					{selectedTimeline === 'trial' ? (
+						<TrialTimeline onSelectLicense={handleLicenseSelect} />
+					) : (
+						<PaidTimeline product={selectedProduct} />
+					)}
+				</div>
+			) : null}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/LicenseSelector/index.tsx:92]:
```suggestion
			{selectedTimeline && (
				<div className="timeline-container">
					{selectedTimeline === 'trial' ? (
						<TrialTimeline onSelectLicense={handleLicenseSelect} />
					) : (
						<PaidTimeline product={selectedProduct} />
					)}
				</div>
			)}
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+	return (
+		<div className="paid-timeline">
+			<div>
+				<div>
+					<span className="mt-3">
+						<p>Need help with license calculations?</p>
+					</span>
+
+					{purchasebleSkus?.map((sku) => {
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx:34]:
```suggestion
					{purchasebleSkus?.map((sku, index) => {
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+						const skuDescription = sku.customFields?.filter(
+							(customFields: any) =>
+								customFields.name === 'Version Description'
+						)[0].customValue.data;
+
+						return (
+							<div className="mb-5" key={sku.sku}>
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx:41]:
```suggestion
							<div className="mb-5" key={index}>
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss [hunk 12]
```diff
@@ -0,0 +1,34 @@
+.trial-timeline {
+	.trial-info {
+		display: flex;
+		color: var(--neutral-08, #54555f);
+		font-size: 16px;
+		font-style: normal;
+		font-weight: 400;
+		line-height: 24px;
+		gap: 4px;
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss:9]: Order should be ascending

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+	.trial-card-icon {
+		width: 16px;
+		height: 16px;
+	}
+
+	.card-button {
+		width: 100%;
+		height: auto;
+	}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss:33]: Order should be ascending

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+						className="trial-card-icon"
+						icon="percentage-symbol"
+					/>
+				}
+				iconRight={true}
+				onClick={() => {
+					setHandleSelected(true);
+					onSelectLicense(true);
+				}}
+				selected={handleSelected ? true : false}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx:43]:
```suggestion
				selected={handleSelected}
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+			<CardButton
+				description="Trial licenses are intended for you to try the app before you buy. Typical trials are 30 days."
+				disabled={false}
+				icon={
+					<GetClayIcon
+						className="trial-card-icon"
+						icon="percentage-symbol"
+					/>
+				}
+				iconRight={true}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx:38]:
```suggestion
				iconRight
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+					<GetClayIcon icon="question-circle-full" />
+				</span>
+			</p>
+			<CardButton
+				description="Trial licenses are intended for you to try the app before you buy. Typical trials are 30 days."
+				disabled={false}
+				icon={
+					<GetClayIcon
+						className="trial-card-icon"
+						icon="percentage-symbol"
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx:35]: I suggested at beginning of review to remove this component and call directly here

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/utils/calculateTrialTime.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+	date.setTime(date.getTime() + 30 * 24 * 60 * 60 * 1000);
+
+	const endOfTrialDay = date.getDate();
+	const endOfTrialMonth = months[date.getMonth()];
+
+	return {
+		endOfTrialDay,
+		endOfTrialMonth,
+	};
+}
```
Comment (kevenleone) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/utils/calculateTrialTime.ts:20]: Avoid using magic numbers, what `30 * 24 * 60 * 60 * 1000` means?

See this example: https://github.com/ryanmcdermott/clean-code-javascript?tab=readme-ov-file#use-searchable-names

```suggestion
import {months} from './getMonths';

CONST ASSIGN_A_BETTER_NAME_TO_ME = 30 * 24 * 60 * 60 * 1000;

export function calculateTrialTime() {
	const date = new Date();

	date.setTime(date.getTime() + ASSIGN_A_BETTER_NAME_TO_ME);

	const endOfTrialDay = date.getDate();
	const endOfTrialMonth = months[date.getMonth()];

	return {
		endOfTrialDay,
		endOfTrialMonth,
	};
}
```

**Review by kevenleone — CHANGES_REQUESTED** [review]
Hi @juliocezar96 , good job on that, I left some suggestions for you

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/utils/calculateTrialTime.ts (hunk 17 above)
Comment (ocesar9) [reply workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/utils/calculateTrialTime.ts:20]: Hi @kevenleone this was not developed by me, I just replaced this function in this file, but will be fixed!

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/GetAppPage.tsx [hunk 18]
```diff
@@ -77,10 +95,124 @@ const GetAppFlow = () => {
...
+		});
+
+		return data;
+	};
+
+	const onsubmit = async (
+		account: Account | undefined,
+		productChannels: Channel | undefined,
+		productSku: SKU | undefined,
+		projectOrderType: ProjectOrderType | undefined
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/GetAppPage.tsx:152]:
```suggestion
		account?: Account,
		productChannels?: Channel,
		productSku?: SKU,
		projectOrderType?: ProjectOrderType
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+interface PaidTimelineProps {
+	product?: Product;
+}
+
+const mock = {
+	licensetiers: [
+		{quantity: 1, value: '$979.99'},
+		{quantity: 3, value: '$900.00'},
+	],
+};
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/PaidTimeline/index.tsx:19]: Do we really need this hard code?

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss [hunk 20]
```diff
@@ -0,0 +1,34 @@
+.trial-timeline {
+	.trial-info {
+		display: flex;
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss:3]: Please use Bootstrap here

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+interface TrialTimelineProps {
+	onSelectLicense: (licenseSelected: boolean) => void;
+}
+
+export function TrialTimeline({onSelectLicense}: TrialTimelineProps) {
+	const [handleSelected, setHandleSelected] = useState<Boolean>(false);
+
+	return (
+		<div className="d-flex flex-column trial-timeline">
+			<p className="mb-2 trial-info">
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx:22]:
```suggestion
			<p className="mb-2 trial-info d-flex">
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+		color: var(--neutral-08, #54555f);
+		font-size: 16px;
+		font-style: normal;
+		font-weight: 400;
+		line-height: 24px;
+		gap: 4px;
+
+		.info-button {
+			display: flex;
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss:12]: Please use Bootstrap here

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+				Need help with license calculations?
+				<span className="info-button">
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.tsx:24]:
```suggestion
				<span className="info-button d-flex">
```

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss (hunk 13 above)
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/src/pages/GetAppPage/components/TrialTimeline/index.scss:33]: Seems like we're not using it

File: workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/yarn.lock [hunk 24]
```diff
@@ -0,0 +1,2895 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
```
Comment (joanagrego) [inline workspaces/liferay-marketplace-workspace/client-extensions/liferay-marketplace-custom-element/yarn.lock:3]: Please remove this file from your commits.

**Review by joanagrego — CHANGES_REQUESTED** [review]

**Comment by juliocezar96** [comment]
Just starting reviewing :)

**Comment by juliocezar96** [comment]
resend #4868

**Pull request closed without merging**